cpp/test/Slice/errorDetection/IdentAsKeyword.err

```diff
@@ -64,20 +64,4 @@ IdentAsKeyword.ice:74: `i8': an interface can only be defined at module scope
 IdentAsKeyword.ice:76: `i9': an interface can only be defined at module scope
 IdentAsKeyword.ice:77: `i10': an interface can only be defined at module scope
 IdentAsKeyword.ice:79: `true': an interface can only be defined at module scope
-IdentAsKeyword.ice:81: illegal leading underscore in identifier `_a'
-IdentAsKeyword.ice:81: `_a': an interface can only be defined at module scope
-IdentAsKeyword.ice:82: illegal leading underscore in identifier `_true'
-IdentAsKeyword.ice:82: `_true': an interface can only be defined at module scope
-IdentAsKeyword.ice:83: illegal leading underscore in identifier `_true'
-IdentAsKeyword.ice:83: `_true': an interface can only be defined at module scope
-IdentAsKeyword.ice:85: illegal trailing underscore in identifier `b_'
-IdentAsKeyword.ice:85: `b_': an interface can only be defined at module scope
-IdentAsKeyword.ice:87: illegal double underscore in identifier `b__c'
-IdentAsKeyword.ice:87: `b__c': an interface can only be defined at module scope
-IdentAsKeyword.ice:88: illegal double underscore in identifier `b___c'
-IdentAsKeyword.ice:88: `b___c': an interface can only be defined at module scope
-IdentAsKeyword.ice:90: illegal underscore in identifier `a_b'
-IdentAsKeyword.ice:90: `a_b': an interface can only be defined at module scope
-IdentAsKeyword.ice:91: illegal underscore in identifier `a_b_c'
-IdentAsKeyword.ice:91: `a_b_c': an interface can only be defined at module scope
-IdentAsKeyword.ice:93: syntax error
+IdentAsKeyword.ice:81: syntax error
```

Comment:
I would fix the class to use a data member `exception` instead of an operation named `exception`.

Reply:
I'm not entirely sure what you're referencing here? Do you mean the operations `exception` and `EXception` in interfaces `i3` and `i4` or something else?

Reply:
Sounds like I misread it. Where is this syntax error on line 81 coming from? Maybe it's because the file doesn't end with a newline?

Reply:
No, we don't require Slice files to end with a newline, even if they should.
It looks like the file ends with a newline anyways, which is good!

After scouring through the compiler's debug output the issue is:
https://github.com/zeroc-ice/ice/blob/41bb61537a8711ea2bef5ae7a49c0a9118dc9a8c/cpp/test/Slice/errorDetection/IdentAsKeyword.ice#L8
As far as the compiler knows INTERFACE is an identifier, so when it hits it, the compiler treats it as an error, and issues the first `syntax error`:
https://github.com/zeroc-ice/ice/blob/41bb61537a8711ea2bef5ae7a49c0a9118dc9a8c/cpp/test/Slice/errorDetection/IdentAsKeyword.err#L1

Then our rule for error handling, is just keep absorbing tokens until we hit a `;` or `}` then try to recover. So the compiler turns
```
module Test
{
INTERFACE i { void op(); }

exception Void();
...
}
```
into
```
module Test
{
}

exception Void();
...
}
```
Since it hits the `;` at the end of the operation.

The last syntax error at the end is because we absorbed an opening brace in the error handling, so the compiler sees a random unmatched `}` at the end of the file, and issues an error for it.
That's also why we get tons of `X can only be defined at module scope` errors, because the compiler already closed the module they're supposed to be, and everything else is being dumped into global scope.

We could try to improve our error handling to not absorb and ignore braces, and instead try to match them, even in error conditions... But this might make handling errors with the user actually has an extra brace harder. I'm not currently sure...

Reply:
Actually no, it would be better to track braces, even in error conditions. We could still catch things like `module {{` because there's no keyword before the brace.
And I think if they've left off a brace, it's going to break anyways. I'll open an issue to add better brace-tracking (even in error conditions) to the compiler to avoid this.